src/arc.ts

```diff
@@ -393,14 +480,40 @@ async function enableWeb3Provider(): Promise<void> {
     throw new Error("Unable to enable provider");
   }
 
-  if (!await initializeArc(provider)) {
-  // eslint-disable-next-line no-console
-    console.error("Unable to initialize Arc");
-    throw new Error("Unable to initialize Arc");
+  biconomy = getBiconomy(provider);
+  biconomy.onEvent(biconomy.READY, async () => {
+    // eslint-disable-next-line no-console
+    console.debug("Biconomy initialized");
+    // Initialize your dapp here like getting user accounts etc
+    if (!await initializeArc(biconomy)) {
+      // eslint-disable-next-line no-console
+      console.error("Unable to initialize Arc");
+      throw new Error("Unable to initialize Arc");
+    }
+    if (initializedAccount) {
+      const userContract = await biconomyLogin(initializedAccount, options);
+      showUserContractAddress(userContract);
+    }
+  }).onEvent(biconomy.ERROR, async (error: any) => {
+    // eslint-disable-next-line no-console
+    console.error(`Unable to initialize biconmy ${JSON.stringify(error)}`);
+    // Fallback to default provider in case of any error
+    // eslint-disable-next-line require-atomic-updates
+    if (!await initializeArc(provider)) {
+      // eslint-disable-next-line no-console
+      console.error("Unable to initialize Arc");
+      throw new Error("Unable to initialize Arc");
+    }
+    selectedProvider = provider;
```

Comment:
We want selectedProvider to only be set in one place in this function.  I also note this is done asynchronously, outside the flow of `enableWeb3Provider` which is likely to cause big problems.  Same with the calls to `initializeArc`.  All this logic is supposed to be complete before `enableWeb3Provider` returns.  We need to create a promise and await it for these the biconomy events to happen.  And only set selectedProvider at the end when everything is complete.

Reply:
I implemented this change in this PR:  https://github.com/orenyodfat/alchemy/pull/2